src/utils/range-set.js

```diff
@@ -0,0 +1,50 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+// @flow
+
+import type { StartEndRange } from 'firefox-profiler/types';
+
+/**
+ * Canonicalize the list of ranges, by OR'ing nested and overlapping ranges
+ * together so that the resulting list of ranges is a flat list of ranges which
+ * covers the same time values. The resulting list has the following properties:
+ *
+ *  - Sorted by range.start
+ *  - No empty ranges
+ *  - No overlap
+ *  - Adjacent ranges are collapsed into one
+ */
+export function canonicalizeRangeSet(ranges: StartEndRange[]): StartEndRange[] {
+  if (ranges.length === 0) {
+    return [];
+  }
+
+  const sortedRanges = ranges.slice().sort((a, b) => a.start - b.start);
+  let lastCanonRange = { ...sortedRanges[0] };
+  const canonRanges = [lastCanonRange];
+
+  for (let i = 1; i < sortedRanges.length; i++) {
+    const range = sortedRanges[i];
+    if (range.start >= range.end) {
+      // Empty or invalid range, skip.
+      continue;
+    }
```

Comment:
I haven't actually checked, but my guess is there weren't any invalid ranges. But I think empty ranges could happen in theory. Anyway, I just put in the check for invalid ranges so that the function works reliably regardless of input.